### ISSUE TITLE
Windows Executable Version

### ISSUE THREAD

**Issue by mothur-westcott** [torognes/vsearch#170]
Thanks for the great addition to the bioinformatics open source software arena! It would be really nice to have a version available for Windows.  I would be willing to help work towards getting one built if you are interested.

**Comment by colinbrislawn** [comment]
:+1: That would be great. 

- VSEARCH would be a 'drop in replacement' across platforms (like USEARCH already is). 
- Many people use Windows, including people who are new to computational biology. 
- The most cited and second most cited software packages in amplicon analysis could use it.

**Comment by torognes** [comment]
Thanks for the suggestion and the offer to help! I'll try to find a suitable machine to compile on locally. Meanwhile, if anyone could start finding problematic portions of the code that will not compile on Windows, including missing library functions etc, it would be helpful.

**Comment by mothur-westcott** [comment]
I started making some modifications on a fork of version 1.9.10.  vsearch.h includes some ifdefs for windows and also visual studio.  The regex library was an issue, but visual studio has an extension for this if you want to use it.  I also wrote a quick parsing version in the abundance files, with ifdefs for regex.  I also ifdefed for pthreads in case that library is not present. Things got a little "in the weeds" and I think perhaps some of the edits I made could be helpful to you and some you may want to discard.  Here's a link to the fork, https://github.com/mothur/vsearch.

**Comment by torognes** [comment]
Which versions of Windows should we target? Would 64bit Windows 7 and later be ok?

**Comment by xflouris** [comment]
I've tried compiling the Windows fork with a cross-compiler (i686-w64-mingw32-g++) and there is a problem when compiling `align_simd.c`. I'm getting the following error:

```
i686-w64-mingw32-g++ -DHAVE_CONFIG_H -I. -I..    -Wall -Wsign-compare -g -O2 -MT align_simd.o -MD -MP -MF $depbase.Tpo -c -o align_simd.o align_simd.cc
align_simd.cc:98:2: error: ‘__m128i’ does not name a type
  __m128i matrix[32];
```

If I explicitly pass the `-mssse3` to the compiler then it works fine.

I'm not sure if this is a problem of the cross-compiler (I'm using version 4.9.1), but it would be nice if someone can check it with MingW directly on Windows.

**Comment by colinbrislawn** [comment]
> Would 64bit Windows 7 and later be ok?

Yes, fully patched 64bit Windows 7 would be a good target. 
Windows 10 is coming on strong and will be supported into the future, so only supporting 64bit Win 10 is reasonable, through a little limiting in the immediate future.

**Comment by torognes** [comment]
I am working on this issue now. Do you think it would be okay with an executable binary that is made using [mingw-w64](https://mingw-w64.org/), perhaps with a cross compiler, as long as it runs nicely and natively on 64-bit Windows? Or does it need to be made using the Microsoft Visual Studio and C++? I think the former seems simpler from a developer perspective.

**Comment by torognes** [comment]
I have compiled a binary of VSEARCH 2.3.1 for 64-bit Windows now. It is available as a pre-release:

https://github.com/torognes/vsearch/releases/tag/v2.3.1-pre-win

It seems to work okay, but I haven't tested in much. Feedback greatly appreciated!

**Comment by mothur-westcott** [comment]
Thanks for releasing this! Mothur users are very happy to be able to use vsearch's functionality, and are anxiously awaiting the Windows version.  I gave it a test drive and had a little trouble.  The output to the screen said 531 OTUs as did the log file, but in the *.uc file some sequences were assigned to OTUs above that number.  

H	700	254	99.2	+	0	0	254M	GQY1XT001D9OYH;size=1;
	GQY1XT001DORYK;size=2;
H	702	255	99.2	+	0	0	174MD80M	GQY1XT001D22UC;size=1;
	GQY1XT001B4G56;size=2;
...

I have attached the files I used and the output files from vsearch.  Thanks again for all the work you are doing with this project!

[vsearch_Windows_Results.zip](https://github.com/torognes/vsearch/files/643189/vsearch_Windows_Results.zip)

**Comment by torognes** [comment]
Thank you for testing it. I think this problem must be due to issue #207 which was fixed in version 2.3.2. I'll make a release for Windows of version 2.3.4 that should fix this problem.

**Comment by torognes** [comment]
Hi @mothur-westcott : VSEARCH 2.3.4 for Windows is now released:

https://github.com/torognes/vsearch/releases/tag/v2.3.4-pre-win

That should fix the problem with the uc file.

**Comment by mothur-westcott** [comment]
Thanks, the fix works great!

**Comment by torognes** [comment]
Included in version 2.4.0 just released.
